rtt/scripting/CallFunction.hpp

```diff
@@ -93,36 +94,127 @@ namespace RTT
+                //    following step 1 this does not give other operations the
+                //    chance to execute.
+                //
+                //    @sa analogous to OperationCallerInterface::isSend()
+                //
+                bool isself = mrunner->isSelf();
+                if (isself) {
+                    maccept = true;
+                    this->executeAndDispose();
+                    if ( _foo->inError() ) {
+                        throw false;
+                    } else if ( _foo->isStopped() ) {
+                        return true;
+                    }
+
+                    // The function yielded!
+                    assert( maccept == false );
+                }
+
+                // 1. Enqueue as a message callback (for the callback step)
+                //    ==> mrunner will call executeAndDispose() (see below)
+                //
+                //    If the caller thread is the same as the executor and the
+                //    funtion yields, loop and enqueue it again. So yielding in
+                //    a function called from the same thread only gives other
+                //    enqueued messages (operations) a chance to execute, but
+                //    afterwards the function continues within the same update
+                //    step and without a call to updateHook() in between.
+                //
+                do {
+                    maccept = mrunner->process( this );
+                    if ( !maccept ) return false;
+
+                    // block for the result: foo stopped or in error or yielded
+                    mrunner->waitForMessages(boost::bind(&CallFunction::fooDone, this) );
                     if ( _foo->inError() ) {
                         throw false;
+                    } else if ( _foo->isStopped() ) {
+                        return true;
                     }
+
+                    // The function yielded!
+                    assert( maccept == false );
+                } while(isself);
+
+                // The function yielded and the calling thread is not the same as the executor.
+                assert(!isself);
+
+                // 2. If not yet finished or in error, enqueue as a function (update step)
+                //    ==> mrunner will run _foo until it finishs
+                // Note: There is no guarantee that the function will not already execute again in the same cycle,
+                //       once as a message and once as function. Not really a problem...
+                maccept = mrunner->runFunction( _foo.get() );
+                if ( !maccept ) return false;
+
+                // block for the result: foo stopped or in error
+                mrunner->waitForMessages(boost::bind(&CallFunction::fooDone, this) );
+                if ( _foo->inError() ) {
+                    throw false;
+                } else if ( _foo->isStopped() ) {
                     return true;
                 }
+
```

Comment:
Looong method. Please split out some helpers. A split may be considered using the steps. Ie. `// 0.`  `// 1.`  `// 2.`